src/StdChains.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2 <0.9.0;
+
+pragma experimental ABIEncoderV2;
+
+import "./Vm.sol";
+
+abstract contract StdChains {
+    VmSafe private constant vm = VmSafe(address(uint160(uint256(keccak256("hevm cheat code")))));
+
+    struct Chain {
+        // The chain name, using underscores as the separator to match `foundry.toml` conventions.
+        string name;
+        // The chain's Chain ID.
+        uint256 chainId;
+        // A default RPC endpoint for this chain.
+        // NOTE: This default RPC URL is included for convenience to facilitate quick tests and
+        // experimentation. Do not use this RPC URL for production test suites, CI, or other heavy
+        // usage as you will be throttled and this is a disservice to others who need this endpoint.
+        string rpcUrl;
+    }
+
+    // Maps from a chain's name (matching what's in the `foundry.toml` file) to chain data.
+    mapping(string => Chain) private chains;
+
+    function getChain(string memory name) internal virtual returns (Chain memory) {
```

Comment:
This is kinda nice because we can also have overloads to get chains by chain ID for user's who want to keep the getter more brief. Of course we'd need a second mapping to map from uint to `Chain` instead, but that might be nice to have.

On a related note, wondering if we should call this method `chains` (and rename the mapping) instead of `getChains`, e.g. which to go with of:
- `getChains("mainnet").rpcUrl`
- `chains("mainnet").rpcUrl`

Reply:
Great idea. Left `getChain` because it "gets" it from the config (at least the initial ones). Feel free to change.

There is a problem if two Chains have the same chain ID (e.g. Anvil and Hardhat) - last Chain will be returned. We can print a warning when the user sets another chain with existing chain ID.

Reply:
> There is a problem if two Chains have the same chain ID (e.g. Anvil and Hardhat) - last Chain will be returned. We can print a warning when the user sets another chain with existing chain ID (or remove the functionality completely so that someone doesn't make a mistake and deploy to wrong chain, etc).

This seems ok to me. I don't think there's any risk of deploying to the wrong chain here, since this would really only happen with anvil/hardhat, otherwise no production chains should duplicate chain ID. Worth noting this in the docs/code comments though, just in case someone adds e.g. two mainnet entries where one is Infura RPC and the other is flashbots

Reply:
My concern is that the user can make an error and end up with multiple entries with the same chain ID, in which case last chain will be returned when getting by chain ID. Let's ping someone from Foundry for opinion. @mds1

Reply:
Right, I don't think it's a big risk since chains have different IDs, and the RPC is the source of truth, but mitigating it is probably better anyway. Some options:
1. When adding a chain in `setChain`, remove the `chainId` input arg and fetch it from the RPC by selecting a fork with that URL then switching back to the prior context. This ensures chains always have the correct chain ID set. We should bypass this and set manually in `initialize()` to avoid RPC calls since we are sure that initial config present is correct, e.g. maybe a private `setChain` overload is used in `initialize` and then we expose a different one for testing
2. Remove the default `hardhat` chain. Then update `setChain` to only support a single instance of a chain ID, e.g. if my test calls `setChain("hardhat", "http://127.0.0.1:8545")` this _replaces_ the existing `anvil` chain and emits a warning log

Reply:
I think `getChain(uint256)` is the problem. Unlike mapping keys, chain IDs may not be unique. We want to support multiple endpoints, for example flashbots, as you said. This means chain IDs can't have 1-1 relation with mapping keys, regardless of `setChain` logic. My proposal is to remove `getChain(uint256)` in favor of multiple endpoints.

Reply:
> We want to support multiple endpoints, for example flashbots, as you said.

Hmm, well this was just an idea I had as to when you may when multiple endpoints for a single chain ID. But no one has requested this yet, and if unsupported you can workaround it with repeated `setChain` + `createSelectFork` use. So I think we should only support one entry per chain ID to keep this feature simple/safer and avoid over-engineering it to support a niche feature that we're not even sure anyone has a use case for

Reply:
Remove `hardhat` and add revert to `setChain`, then?

Reply:
I'd say so, yea let's keep anvil as the default and people can always `setChain` to update the name/RPC URL. My suggestion is:
- Use chain ID as the source of truth for a chain
- Map chain IDs to the key names (where key name = the alias in foundry.toml)
- Have mappings from (1) chain ID to `Chain` struct, and (2) from key name to `Chain` struct
- Expose getters to get `Chain` info by either chain ID or key name
- Expose setter that takes key name + RPC URL, selects that fork, fetches chain ID, switches back to original fork/context (not sure if this is currently possible, let's check with @mattsse) (this also serves to validate the RPC URL)
- Setter can both overwrite an existing RPC or set a new one
- Our `initialize` method should _not_ use that setter, because we don't want dozens of chain ID RPC requests on initialization

Sorry for the churn here, but lmk how that sounds to you

Reply:
A mapping from chain ID to `Chain` is redundant. I think the current set up is better: chain ID -> key -> `Chain`.

Reply:
That's fine! I just meant conceptually both should exist in that the user can access `Chain` info from either approach.

Also it seems there's currently no way to switch back to a no-fork context, so for now let's just keep `setChain(name, key, chainId, rpcUrl)`, and we can add the no-chainId version in the future